### DISCUSSION TITLE
How to test a finalized application built by terminal.gui

### DISCUSSION THREAD

**Discussion by ZerNox** [tui-cs/Terminal.Gui#3185] [General]
Lets consider that we have built the an installer using Terminal.Gui and want to test, and we want to test it in an automated way. 
We have previously used a simple expect sequence to go through the options in a bash script and verify in the backend. 
Expect does not seem to work for Terminal.Gui or do you need to do something special to get it to work?
Or how can this be achieved any advice's or examples where there are tests wrt sequence to go through the options in a bash script and verify in the backend. 
Expect does not seem to work for Terminal.Gui or do you need to do something special to get it to work?
Or how can this be achieved any advice's or examples where there are tests written? 
itten?

**Comment by tznind** [comment]
Are you using v2 or v1? Here is an example of an automation style test for `FileDialog` in which `Application.Driver.SendKeys` is used.

https://github.com/gui-cs/Terminal.Gui/blob/93659510c368c33e5813126fe5116cbe708688d0/UnitTests/FileServices/FileDialogTests.cs#L32-L57

> We have previously used a simple expect sequence to go through the options in a bash script and verify in the backend.
> Expect does not seem to work for Terminal.Gui or do you need to do something special to get it to work?

Is the goal that you treat the app as a 'black box' and have a pre defined series of keypresses/clicks you want executed at a given timings while it is running? Presumably after which you want to look at the environment or process output?

**Reply by ZerNox** [reply]
Thanks @tznind it seems like an acceptable approach but im really new in this field of testing. Do you recommend this way of doing it?

**Reply by tznind** [reply]
Yes, assuming you have the source code to the app and can create (or expand an existing) test project.

If you are wanting to test the app in a non development environment and only have access to the compiled binary then you would need some kind of external key automation tool.

What are you currently using for testing your shell scripts? Something like [pexpect](https://github.com/pexpect/pexpect) may still work with Terminal.Gui applications, I haven't tried.
